src/lib.rs

```diff
@@ -68,6 +113,113 @@ pub struct PortGroup {
 pub struct Port {
     // Path of the domain socket.
     pub socket_path: &'static str,
+    pub socket: UnixStream,
+}
+
+impl Port {
+    pub fn new(path: &'static str) -> Port {
+        // Connect to the unix domain socket for this port
+        let socket = UnixStream::connect(path).unwrap();
+        // Create and return the port struct
+        Port {
+            socket_path: path,
+            socket: socket,
+        }
+    }
+
+    pub fn i2c(&mut self, address: u8, frequency: u32) -> I2C {
+        // Create and return the I2C struct
+        I2C::new(self, address, frequency)
```

Comment:
One caveat I'll note with this kind of API is that there can only be one `&mut` borrow of an object in existence at any given time, so you while you have an I2C object, you aren't allowed to touch the Port directly. That's almost a good thing as it prevents you from enabling SPI + I2C + UART at once, but it does prevent using non-I2C pins as GPIO through the Port object.

I can think of a couple of ways around this:

* To be like the JS API, with no compile-time checks on API misuse, you could have a `Backend` struct that holds the domain socket and manages the command queue. `Port`, `I2C`, and all other objects could share the same backend by holding `Arc<Mutex<Backend>>` and call methods on it after acquiring the lock. The Port and the I2C objects could be sent to separate threads, and the Backend would be responsible at runtime for making sure that the options they're asking for don't conflict with each other.

* You could add methods to the I2C struct to access the GPIO and other functionality that are legal to use in I2C mode. In this case, you'd probably want to make this method take `self` by value, and have I2C own `Port` rather than `&mut Port`. This way would mean separate types representing the port for unconfigured vs I2C vs SPI, etc, with methods to convert between the types that also send commands to change the mode. This would require all code that uses the port to be aware of what mode it's in, which doesn't seem completely unreasonable.